core/unix/memquery_test.h

```diff
@@ -0,0 +1,311 @@
+
+#include "memquery_test_data.h"
+
+memquery_library_bounds_test cur_bounds;
+int next_iter_position = -1;
+bool enable_memquery_unit_test;
+
+/* CHECK_EQ verifies that the provded x and y are equal; if not, it prints a
+ * message to STDERR and returns false.
+ */
+#define CHECK_EQ(x, y, fmt)                                                  \
+    if ((x) != (y)) {                                                        \
+        print_file(STDERR, #x " != " #y ": " fmt " != " fmt "\n", (x), (y)); \
+        return false;                                                        \
+    }
+
+/* Piggyback on CHECK_EQ for the failure message: since the strncmp
+ * showed the strings are not equal, we know that the x and y pointers are also
+ * not equal according to ==.
+ */
+#define ASSERT_STR_EQ(x, y, sz)         \
+    if (strncmp((x), (y), (sz)) != 0) { \
+        CHECK_EQ(x, y, "%s");           \
+    }
+
+bool
+run_single_memquery_test(memquery_library_bounds_test test)
+{
+    cur_bounds = test;
+
+    char fulldir[200];
+    char filename[200];
+
+    app_pc start = test.in_start;
+    app_pc end;
+
+    int got_return = memquery_library_bounds_by_iterator(
+        test.in_name, &start, &end, test.want_fulldir ? fulldir : NULL,
+        BUFFER_SIZE_ELEMENTS(fulldir), test.want_filename ? filename : NULL,
+        BUFFER_SIZE_ELEMENTS(filename));
+
+    CHECK_EQ(start, test.want_start, PFX);
+    CHECK_EQ(end, test.want_end, PFX);
+    CHECK_EQ(got_return, test.want_return, "%d");
+    if (test.want_fulldir) {
+        ASSERT_STR_EQ(fulldir, test.want_fulldir, BUFFER_SIZE_ELEMENTS(fulldir));
+    }
+    if (test.want_filename) {
+        ASSERT_STR_EQ(filename, test.want_filename, BUFFER_SIZE_ELEMENTS(filename));
+    }
+    return true;
+}
+
+void
+unit_test_memquery()
+{
+    bool all_passed = true;
+    print_file(STDERR, "START unit_test_memquery\n");
+    enable_memquery_unit_test = true;
+    for (int i = 0; i < NUM_MEMQUERY_TESTS; i++) {
+        const char *test_name = all_memquery_test_names[i];
+        print_file(STDERR, "Run memquery unit test %s...\n", test_name);
+        const char *status = "OK";
+        if (!run_single_memquery_test(*all_memquery_tests[i])) {
+            status = "FAILED";
+            all_passed = false;
+        }
+        print_file(STDERR, "**** memquery unit test %s %s\n", test_name, status);
+    }
+    enable_memquery_unit_test = false;
+    if (!all_passed) {
+        print_file(STDERR, "unit_test_memquery had failing test(s)!\n");
+        os_terminate(NULL, TERMINATE_PROCESS);
+    }
+    print_file(STDERR, "END unit_test_memquery\n");
+}
+
+bool
+memquery_iterator_start_test(memquery_iter_t *iter, app_pc start, bool may_alloc)
+{
+    if (!enable_memquery_unit_test) {
+        return memquery_iterator_start(iter, start, may_alloc);
+    }
+    next_iter_position = 0;
+    return true;
+}
+
+bool
+memquery_iterator_next_test(memquery_iter_t *iter)
+{
+    if (!enable_memquery_unit_test) {
```

Comment:
I also guessed that it wasn't needed initially, but when I first ran the test without this bool (for a single simple test case) it failed somewhere in os.c during some routine that I did not recognize; I didn't look too hard since introducing this bool got things working again, I just assumed it was expected to use some of these routines at some point when other arbitrary core code was running.

I believe we can remove this flag if we move to a more strict test API anyway, which is nice because this felt a little gross.